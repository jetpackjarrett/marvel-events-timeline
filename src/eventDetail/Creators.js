import React, { PropTypes } from 'react';

const Creators = ({ creators }) => {
  const creatorsByRole = creators.reduce((memo, { role, name }) => {
    // the Marvel API often returns penciller missing an l
    if (role === 'penciler') {
      role = 'penciller';
    }

    if (!memo[role]) {
      memo[role] = [];
    }

    memo[role].push(name);
    return memo;
  }, {});
  const roles = Object.keys(creatorsByRole).sort();

  return (
    <section>
      <h4>Creators</h4>
      <ul className="event-detail-creators">
        {roles.map((role) => {
          return (
            <li key={role} className="event-detail-creator">
              <strong>{creatorsByRole[role].length > 1 ? `${role}s` : role}</strong>
              <span>{creatorsByRole[role].join(', ')}</span>
            </li>
          );
        })}
      </ul>
    </section>
  );
};

Creators.propTypes = {
  creators: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    role: PropTypes.string.isRequired,
  })).isRequired,
};

export default Creators;
