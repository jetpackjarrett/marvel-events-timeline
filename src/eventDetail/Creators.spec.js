/*eslint import/no-extraneous-dependencies: ["error", { "devDependencies": true }] */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Creators from './Creators';

describe('EventDetail Creators', () => {
  it('renders the creators grouped by role', () => {
    const component = shallow(
      <Creators
        creators={[
          {
            name: 'Stan Lee',
            role: 'writer'
          },
          {
            name: 'Jack Kirby',
            role: 'penciller',
          },
          {
            name: 'Steve Ditko',
            role: 'penciller',
          },
          {
            name: 'John Romita Sr.',
            role: 'penciler',
          }
        ]} />
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});
