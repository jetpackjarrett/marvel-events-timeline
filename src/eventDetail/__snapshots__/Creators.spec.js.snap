exports[`EventDetail Creators renders the creators grouped by role 1`] = `
<section>
  <h4>
    Creators
  </h4>
  <ul
    className="event-detail-creators">
    <li
      className="event-detail-creator">
      <strong>
        pencillers
      </strong>
      <span>
        Jack Kirby, Steve Ditko, John Romita Sr.
      </span>
    </li>
    <li
      className="event-detail-creator">
      <strong>
        writer
      </strong>
      <span>
        Stan Lee
      </span>
    </li>
  </ul>
</section>
`;
